src/gl/tessellation_helpers.cpp

```diff
@@ -14,32 +14,58 @@
  * along with this program. If not, see <http://www.gnu.org/licenses/>.
  */
 #include "mir/gl/tessellation_helpers.h"
+#include "mir/geometry/forward.h"
 #include "mir/graphics/renderable.h"
 #include "mir/graphics/buffer.h"
 
 namespace mg = mir::graphics;
 namespace mgl = mir::gl;
 namespace geom = mir::geometry;
+
+namespace
+{
+struct SrcTexCoords
+{
+    GLfloat top;
+    GLfloat bottom;
+    GLfloat left;
+    GLfloat right;
+};
+
+auto tex_coords_from_rect(geom::Size buffer_size, geom::RectangleD sample_rect) -> SrcTexCoords
+{
+    /* GL Texture coordinates are normalised to the size of the buffer, so (0.0, 0.0) is the top-left
+     * and (1.0, 1.0) is the bottom-right
+     */
+    SrcTexCoords coords;
+    coords.top = sample_rect.top().as_value() / buffer_size.height.as_uint32_t();
+    coords.bottom = sample_rect.bottom().as_value() / buffer_size.height.as_uint32_t();
+    coords.left = sample_rect.left().as_value() / buffer_size.width.as_uint32_t();
+    coords.right = sample_rect.right().as_value() / buffer_size.width.as_uint32_t();
```

Comment:
Hmm. All this `.as_value()` ... `.as_uint32_t()` makes me wonder if there's an operator missing in our geometry. This would look safer:

```c++
coords.top = sample_rect.top() / buffer_size;
```

Reply:
I've ended up with `sample_rect.top() / buffer_size.width` (as `sample_rect.top() / buffer_size` is ambiguous).